### PR TITLE
make tap symlink free

### PR DESCRIPTION
**Pull request by xu-cheng**
Per https://github.com/Homebrew/homebrew/issues/36731

Ping @mikemcquaid, @jacknagel, @DomT4

### CODE REVIEW

File: Library/Homebrew/cmd/update.rb [hunk 1]
```diff
@@ -65,6 +55,9 @@ def update
       install_tap tap_user, tap_repo
     end if load_tap_migrations
 
+    # migrate to new directories based tap struct
+    migrate_taps
```
Comment (xu-cheng) [inline Library/Homebrew/cmd/update.rb]: Auto migrate to new way to manage tap formulae.

File: Library/Homebrew/formula.rb [hunk 2]
```diff
@@ -553,13 +562,33 @@ def print_tap_action options={}
...
   # True if this formula is provided by Homebrew itself
   def core_formula?
-    path == Formula.path(name)
+    path == Formula.core_path(name)
   end
 
-  def self.path name
+  def self.core_path name
     Pathname.new("#{HOMEBREW_LIBRARY}/Formula/#{name.downcase}.rb")
   end
```
Comment (xu-cheng) [inline Library/Homebrew/formula.rb]: I rename the old `path` method to `core_path`.

File: Library/Homebrew/formula.rb [hunk 3]
```diff
@@ -553,13 +562,33 @@ def print_tap_action options={}
...
+    else
+      ofail <<-EOS.undent
+        Multiple formulae found:
+        * #{p.map(&:to_s).join("\n        * ")}
+
+        Please use the fully-qualified name to refer the formula.
+      EOS
+    end
+  end
+
```
Comment (xu-cheng) [inline Library/Homebrew/formula.rb]: New `path` method will try to locate formula from core at first. If fails, it will search `Taps` to locate tap formula. There will be three outcome:
1. Found nothing, return `nil` to indicate no formula available.
2. Found one, return it.
3. Found more than one, raise error.

File: Library/Homebrew/formulary.rb [hunk 4]
```diff
@@ -183,6 +183,7 @@ def get_formula(spec)
   class NullLoader < FormulaLoader
     def initialize(name)
       @name = name
+      @path = nil
```
Comment (xu-cheng) [inline Library/Homebrew/formulary.rb]: Let all `FormulaLoader` instances have `path` attribute. And `nil` means formula unavailable.

File: Library/Homebrew/formula.rb [hunk 5]
```diff
@@ -494,7 +494,16 @@ def python(options={}, &block)
...
+    taps = Dir["#{HOMEBREW_LIBRARY}/Taps/*/*/"].reduce([]) do |formulae, tap|
+      Pathname.new(tap).find_formula do |f|
+        if f.to_s =~ HOMEBREW_TAP_PATH_REGEX
+           formulae << "#$1/#{$2.gsub(/^homebrew-/, "")}/#{f.basename(".rb")}"
+        end
+      end
+      formulae
+    end.sort
+    core+taps
   end
```
Comment (xu-cheng) [inline Library/Homebrew/formula.rb]: `Formula.names` will return all formula names. For core formulae, it will be as usual. For tap formulae, return fully-qualified name.

File: Library/Homebrew/cmd/tap.rb [hunk 6]
```diff
@@ -39,59 +38,23 @@ def install_tap user, repo
...
-        ignores << path.basename.to_s
-        tapped += 1
-      end
+  def migrate_taps(autoskip=true)
+    # migrate taps from symlink based to directories based.
+    ignore = HOMEBREW_LIBRARY/"Formula/.gitignore"
+    if ignore.exist?
+      File.delete ignore
+    else
+      return if autoskip
```
Comment (xu-cheng) [inline Library/Homebrew/cmd/tap.rb]: Using `HOMEBREW_LIBRARY/"Formula/.gitignore"` to determine whether we should migrate or not.

File: Library/Homebrew/cmd/tap.rb [hunk 7]
```diff
@@ -39,59 +38,23 @@ def install_tap user, repo
...
-      if not File.exist? fn
-        File.delete fn
-        count += 1
+      f = Pathname.new(fn)
+      next unless f.symlink?
+      if f.exist?
+        File.delete ignore if f.realpath.to_s.start_with? "#{HOMEBREW_LIBRARY}/Taps"
+      else
+        File.delete ignore
       end
```
Comment (xu-cheng) [inline Library/Homebrew/cmd/tap.rb]: Only delete dead symlinks and symlinks from Taps. If users create their own symlinks, they will remain intact.

File: Library/Homebrew/cmd/tap.rb [hunk 8]
```diff
@@ -39,59 +38,23 @@ def install_tap user, repo
...
-          Could not create link for #{Tty.white}#{tap_ref(path)}#{Tty.reset}, as it
-          conflicts with #{Tty.white}#{tap_ref(to)}#{Tty.reset}. You will need to use the
-          fully-qualified name when referring this formula, e.g.
-            brew install #{tap_ref(path)}
-          EOS
-      else
-        ignores << path.basename.to_s
-        tapped += 1
-      end
+  def migrate_taps(autoskip=true)
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/tap.rb]:
```ruby
def migrate_taps(options = { :autoskip => true })
```
Makes calls to the function a bit more readable.

File: Library/Homebrew/formula.rb (hunk 5 above)
Comment (mikemcquaid) [inline Library/Homebrew/formula.rb]: I'd consider not returning the fully-qualified name here for taps and perhaps just add another method that does that. I think, despite it being a private API, that changing the output format for taps here will break some e.g. external commands.

File: Library/Homebrew/formula.rb (hunk 2 above)
Comment (mikemcquaid) [inline Library/Homebrew/formula.rb]: Let's do it the other way round; make the new method have the new functionality.

File: Library/Homebrew/formula.rb (hunk 3 above)
Comment (mikemcquaid) [inline Library/Homebrew/formula.rb]: :+1: 

Just something to consider for the future: we might want to provide a way for people to prioritise taps over others and even ignore the formulae in core altogether.

**Comment by mikemcquaid** [comment]
I'm impressed how quickly you've got something working here. Will want other @Homebrew/owners input and work on the above comments but it's looking pretty good so far. Nice work :+1:

File: Library/Homebrew/formula.rb (hunk 2 above)
Comment (xu-cheng) [inline Library/Homebrew/formula.rb]: OK.

File: Library/Homebrew/formula.rb (hunk 3 above)
Comment (xu-cheng) [inline Library/Homebrew/formula.rb]: This doesn't affect fully-qualified name and load formula using path.

File: Library/Homebrew/cmd/tap.rb (hunk 8 above)
Comment (xu-cheng) [inline Library/Homebrew/cmd/tap.rb]: OK

File: Library/Homebrew/formula.rb (hunk 5 above)
Comment (xu-cheng) [inline Library/Homebrew/formula.rb]: Since we drop the symlink, it also means formula whose name conflicts with core can be allowed. Therefore, using fully-qualified name to present tap formulae is a nature way and better way to do. Otherwise, it would be strange to return some of tapped formulae (that is we cannot predict which formula will return if some of them have the same name). On the other hand, passing fully-qualified name to caller can avoiding further path searching. It's good in term of performance.

For external commands, AFAIK, `homebrew-anytap` is broken (this is obvious) and `homebrew-livechek` failed to handle that path can be nil. I will report the issues to these projects after we are ready.

If we will make sure we don't break any external commands. How about only returning core formulae? Like this:
```ruby
def self.names options={:includetap => false}
   ...
  return core unless options[:includetap]
   ...
end
```

Comment (mikemcquaid) [inline Library/Homebrew/formula.rb]: Well, if we return all the names of the formulae across all taps then it'll be consistent with the current behaviour. It doesn't really matter hugely if some of those names are duplicated across taps for this first pass, I think.

**Comment by xu-cheng** [comment]
@mikemcquaid I have updated to follow all of your suggestions.

File: Library/Homebrew/formula.rb [hunk 9]
```diff
@@ -560,6 +573,25 @@ def self.path name
...
+    return p if p.file?
+    name = name.downcase
+    taps = Dir["#{HOMEBREW_LIBRARY}/Taps/*/*/"]
+    p = taps.map { |tap| Dir["#{tap}#{name}.rb", "#{tap}Formula/#{name}.rb", "#{tap}HomebrewFormula/#{name}.rb"] }
+    p = p.flatten.map(&Pathname.method(:new)).select { |f| f.file? }
+    case p.size
+    when 0; nil
+    when 1; p[0]
+    else
+      odie <<-EOS.undent
```
Comment (xu-cheng) [inline Library/Homebrew/formula.rb]: I'm not sure here. Should I use `odie` or `ofail` + `raise` or something else? @mikemcquaid

File: Library/Homebrew/formula.rb [hunk 10]
```diff
@@ -493,8 +493,21 @@ def python(options={}, &block)
...
+            formulae << "#$1/#{$2.gsub(/^homebrew-/, "")}/#{f.basename(".rb")}"
+          end
+        else
+          formulae << f.basename(".rb").to_s
+        end
+      end
+      formulae
+    end.sort
+    (core+taps).uniq
   end
```
Comment (mikemcquaid) [inline Library/Homebrew/formula.rb]: Again, I'd rather this function behaved as it did before and we add a new function for returning fully-qualified names.

Comment (xu-cheng) [inline Library/Homebrew/formula.rb]: It is behaving as before. The default option is disable full name.

File: Library/Homebrew/formula.rb (hunk 9 above)
Comment (mikemcquaid) [inline Library/Homebrew/formula.rb]: I'm not sure, will let @jacknagel chip in.

File: Library/Homebrew/formula.rb (hunk 10 above)
Comment (mikemcquaid) [inline Library/Homebrew/formula.rb]: Sorry, misread, thanks.

File: Library/Homebrew/formula.rb [hunk 11]
```diff
@@ -560,6 +573,25 @@ def self.path name
     Pathname.new("#{HOMEBREW_LIBRARY}/Formula/#{name.downcase}.rb")
   end
 
+  def self.search_path name
+    p = Formula.path(name)
+    return p if p.file?
+    name = name.downcase
+    taps = Dir["#{HOMEBREW_LIBRARY}/Taps/*/*/"]
+    p = taps.map { |tap| Dir["#{tap}#{name}.rb", "#{tap}Formula/#{name}.rb", "#{tap}HomebrewFormula/#{name}.rb"] }
+    p = p.flatten.map(&Pathname.method(:new)).select { |f| f.file? }
```
Comment (jacknagel) [inline Library/Homebrew/formula.rb]: Just use a regular block here, please, or use `Pathname.glob` instead of `Dir[]`

File: Library/Homebrew/formula.rb (hunk 3 above)
Comment (jacknagel) [inline Library/Homebrew/formula.rb]: Why not raise FormulaUnavailableError instead of returning nil?

File: Library/Homebrew/cmd/tap.rb [hunk 12]
```diff
@@ -39,59 +38,23 @@ def install_tap user, repo
...
-  end
-
-  def repair_taps(warn_about_conflicts=true)
-    count = 0
-    # prune dead symlinks in Formula
     Dir.glob("#{HOMEBREW_LIBRARY}/Formula/*.rb") do |fn|
-      if not File.exist? fn
-        File.delete fn
-        count += 1
+      f = Pathname.new(fn)
```
Comment (jacknagel) [inline Library/Homebrew/cmd/tap.rb]: This can be accomplished by using `Pathname.glob` instead of `Dir.glob`

File: Library/Homebrew/cmd/tap.rb [hunk 13]
```diff
@@ -39,59 +38,23 @@ def install_tap user, repo
...
+      next unless f.symlink?
+      if f.exist?
+        File.delete ignore if f.realpath.to_s.start_with? "#{HOMEBREW_LIBRARY}/Taps"
```
Comment (jacknagel) [inline Library/Homebrew/cmd/tap.rb]: `realpath` is something I've been trying to remove from the codebase where it isn't needed. There should only be one symlink, right? Then we can use `resolved_path`.

File: Library/Homebrew/cmd/update.rb [hunk 14]
```diff
@@ -95,7 +88,6 @@ def git_init_if_necessary
   end
 
   def rename_taps_dir_if_necessary
-    need_repair_taps = false
```
Comment (jacknagel) [inline Library/Homebrew/cmd/update.rb]: Why was this removed?

File: Library/Homebrew/formula.rb (hunk 3 above)
Comment (xu-cheng) [inline Library/Homebrew/formula.rb]: Because, this method is mainly called by `Formulary.loader_for`. If the formula is not found here, it will try to find from Alias and cache.

File: Library/Homebrew/cmd/update.rb (hunk 14 above)
Comment (xu-cheng) [inline Library/Homebrew/cmd/update.rb]: Now symlink is gone, there is no broken tap and such thing as repairing tap.

File: Library/Homebrew/cmd/edit.rb [hunk 15]
```diff
@@ -26,13 +26,17 @@ def edit
     else
       # Don't use ARGV.formulae as that will throw if the file doesn't parse
       paths = ARGV.named.map do |name|
-        name = Formulary.canonical_name(name)
-        Formula.path(name)
-      end
-      unless ARGV.force?
-        paths.each do |path|
-          raise FormulaUnavailableError, path.basename('.rb').to_s unless path.file?
+        formula = Formulary.loader_for(name)
```
Comment (jacknagel) [inline Library/Homebrew/cmd/edit.rb]: I don't want to start using loader_for like this, it shouldn't be used outside of `formulary.rb`.

**Comment by jacknagel** [comment]
A few other notes:

 * I dislike methods that alter their behavior based on a boolean argument. This is a design smell. Can we extract the common bits to a private method, and have multiple methods instead?

 * I dislike making nil "mean something" as a design choice. Can we eliminate the nil somehow, perhaps using polymorphism?

File: Library/Homebrew/cmd/edit.rb (hunk 15 above)
Comment (xu-cheng) [inline Library/Homebrew/cmd/edit.rb]: I'm trying to reduce the IO operations.

Comment (jacknagel) [inline Library/Homebrew/cmd/edit.rb]: Premature optimization.

File: Library/Homebrew/cmd/log.rb [hunk 16]
```diff
@@ -1,15 +1,15 @@
...
       cd HOMEBREW_REPOSITORY
       exec "git", "log", *ARGV.options_only
     else
-      begin
-        path = ARGV.formulae.first.path
-      rescue FormulaUnavailableError
-        # Maybe the formula was deleted
-        path = Formula.path(ARGV.named.first)
-      end
+      path = Formulary.loader_for(ARGV.named.first).path
```
Comment (jacknagel) [inline Library/Homebrew/cmd/log.rb]: Same thing here. I don't see a good reason to regress the design in this way.

Comment (xu-cheng) [inline Library/Homebrew/cmd/log.rb]: This place is a little different. Original design will have weird response if I do something like this: 
`brew log foo bar` if foo is in the tap and bar doesn't exist. It will assume foo is in the core.

**Comment by xu-cheng** [comment]
Updated a bit. Will try to find a better way to eliminate `nil`.

**Comment by xu-cheng** [comment]
Now I have drop the `nil` instead to use exceptions. To make things clearer and easier, I make sure a FormulaLoader either have a existed path or raise an error. There wouldn't be a state that I get path but the file doesn't exist.

Also I add multiple new methods to hide internal implementations.

File: Library/Homebrew/cmd/log.rb [hunk 17]
```diff
@@ -5,9 +8,11 @@ def log
       exec "git", "log", *ARGV.options_only
     else
       begin
-        path = ARGV.formulae.first.path
+        path = Formulary.canonical_path(ARGV.named.first)
+      # Maybe the formula was deleted
+      rescue TapFormulaUnavailableError => e
+        path = HOMEBREW_LIBRARY/"Taps/#{e.user}/homebrew-#{e.repo}/#{e.shortname}.rb"
```
Comment (xu-cheng) [inline Library/Homebrew/cmd/log.rb]: Should I create a method like `Formula.tap_path`?

File: Library/Homebrew/formulary.rb [hunk 18]
```diff
@@ -204,6 +208,15 @@ def self.canonical_name(ref)
...
 
+  def self.canonical_name_and_path(ref)
+    formula = loader_for(ref)
+    [formula.name, formula.path]
+  end
+
+  def self.canonical_path(ref)
+    loader_for(ref).path
+  end
+
```
Comment (xu-cheng) [inline Library/Homebrew/formulary.rb]: new methods to shade internal implementations which can be useful.

File: Library/Homebrew/formulary.rb [hunk 19]
```diff
@@ -160,17 +160,17 @@ def initialize tapped_name
...
-            path = file
+            super name, file
+            return
           end
         end
       end
 
-      super name, path
+      raise TapFormulaUnavailableError, tapped_name
     end
```
Comment (xu-cheng) [inline Library/Homebrew/formulary.rb]: Make sure either returning the founded tapped formula or raising error. There will be no state that to return the loader but the path doesn't exist. Therefore, we guarantee  that path always existing to downstream.

**Comment by xu-cheng** [comment]
Ask for reviewing :smiley:

**Comment by asparagui** [comment]
nice! :tropical_fish:

File: Library/Homebrew/cmd/tap.rb [hunk 20]
```diff
@@ -39,59 +38,22 @@ def install_tap user, repo
...
-    Dir.glob("#{HOMEBREW_LIBRARY}/Formula/*.rb") do |fn|
-      if not File.exist? fn
-        File.delete fn
-        count += 1
+    Pathname.glob("#{HOMEBREW_LIBRARY}/Formula/*.rb") do |f|
+      next unless f.symlink?
+      if f.exist?
+        File.delete ignore if f.resolved_path.to_s.start_with? "#{HOMEBREW_LIBRARY}/Taps"
+      else
+        File.delete ignore
```
Comment (tdsmith) [inline Library/Homebrew/cmd/tap.rb]: Do you want to delete `f` (not `ignore`) in these two lines here?

Comment (xu-cheng) [inline Library/Homebrew/cmd/tap.rb]: Thanks for catching this. Fixed.

File: Library/Homebrew/cmd/install.rb [hunk 21]
```diff
@@ -15,7 +15,9 @@ def install
 
     ARGV.named.each do |name|
       # if a formula has been tapped ignore the blacklisting
-      unless Formula.path(name).file?
+      begin
+        Formula.search_path(name)
+      rescue FormulaUnavailableError
         msg = blacklisted? name
         raise "No available formula for #{name}\n#{msg}" if msg
       end
```
Comment (xu-cheng) [inline Library/Homebrew/cmd/install.rb]: This whole block is kind of weird. I will try to refactor it.

**Comment by xu-cheng** [comment]
I did some refactors on the `install.rb` as the original code doesn't make sense to me.

File: Library/Homebrew/cmd/install.rb [hunk 22]
```diff
@@ -83,12 +77,17 @@ def install
...
-      ofail e.message
-      query = query_regexp(e.name)
-      puts 'Searching formulae...'
-      puts_columns(search_formulae(query))
-      puts 'Searching taps...'
-      puts_columns(search_taps(query))
+      msg = blacklisted? e.name
+      if msg
+        ofail "#{e.message}\n#{msg}"
+      else
```
Comment (xu-cheng) [inline Library/Homebrew/cmd/install.rb]: Move blacklisted check here. Because we only show blacklist message when the formula is not available and blacklisted. Right?

File: Library/Homebrew/cmd/install.rb [hunk 23]
```diff
@@ -14,16 +14,10 @@ def install
...
     ARGV.named.each do |name|
-      # if a formula has been tapped ignore the blacklisting
-      unless Formula.path(name).file?
-        msg = blacklisted? name
-        raise "No available formula for #{name}\n#{msg}" if msg
+      if name =~ HOMEBREW_TAP_FORMULA_REGEX || name =~ HOMEBREW_CASK_TAP_FORMULA_REGEX
+        install_tap $1, $2 unless (HOMEBREW_LIBRARY/"Taps/#{$1.downcase}/homebrew-#{$2.downcase}").directory?
       end
-      if !File.exist?(name) && (name =~ HOMEBREW_TAP_FORMULA_REGEX \
-                                || name =~ HOMEBREW_CASK_TAP_FORMULA_REGEX)
```
Comment (xu-cheng) [inline Library/Homebrew/cmd/install.rb]: The `File.exist?` checked nothing.

File: Library/Homebrew/cmd/install.rb [hunk 24]
```diff
@@ -14,16 +14,10 @@ def install
...
-        install_tap $1, $2
-      end
-    end unless ARGV.force?
```
Comment (xu-cheng) [inline Library/Homebrew/cmd/install.rb]: `ARGV.force?` has no meaning. You want to force to install an unavailable formula? Make no sense.

File: Library/Homebrew/cmd/install.rb (hunk 22 above)
Comment (mikemcquaid) [inline Library/Homebrew/cmd/install.rb]: I believe we previously showed it even if it was available.

Comment (xu-cheng) [inline Library/Homebrew/cmd/install.rb]: Nope.

> # if a formula has been tapped ignore the blacklisting

Comment (mikemcquaid) [inline Library/Homebrew/cmd/install.rb]: Sorry, I mean previously as in: possibly a very long time ago :grinning:

File: Library/Homebrew/cmd/install.rb (hunk 24 above)
Comment (jacknagel) [inline Library/Homebrew/cmd/install.rb]: It's so you can override the blacklist using `--force`.

File: Library/Homebrew/cmd/install.rb (hunk 23 above)
Comment (jacknagel) [inline Library/Homebrew/cmd/install.rb]: I'm pretty sure it had something to do with installing from a path (e.g. `brew install foo/bar/baz.rb`) matching the tap regex.

File: Library/Homebrew/cmd/install.rb (hunk 24 above)
Comment (xu-cheng) [inline Library/Homebrew/cmd/install.rb]: The blacklist will only kick in if the formula cannot be found. So what's the difference using `--force`?

File: Library/Homebrew/formulary.rb (hunk 19 above)
Comment (jacknagel) [inline Library/Homebrew/formulary.rb]: I'm not sure I understand the reason for this change. Can't we just let get_formula be the only place that raises the exception? It's simpler.

File: Library/Homebrew/cmd/log.rb (hunk 17 above)
Comment (jacknagel) [inline Library/Homebrew/cmd/log.rb]: Can we add the path to the exception object where it is raised, instead of reconstructing it here?

File: Library/Homebrew/cmd/install.rb (hunk 23 above)
Comment (xu-cheng) [inline Library/Homebrew/cmd/install.rb]: Oh, I will add it back.

File: Library/Homebrew/formula.rb [hunk 25]
```diff
@@ -560,6 +565,20 @@ def self.path name
...
+  def self.search_path name
+    p = Formula.path(name)
+    return p if p.file?
+    name = name.downcase
+    taps = Dir["#{HOMEBREW_LIBRARY}/Taps/*/*/"]
+    p = taps.map { |tap| Pathname.glob(["#{tap}#{name}.rb", "#{tap}Formula/#{name}.rb", "#{tap}HomebrewFormula/#{name}.rb"]) }
+    p = p.flatten.select { |f| f.file? }
+    case p.size
+    when 0; raise FormulaUnavailableError.new(name)
+    when 1; p[0]
```
Comment (jacknagel) [inline Library/Homebrew/formula.rb]: Use `then` instead of `;`

File: Library/Homebrew/formula.rb [hunk 26]
```diff
@@ -560,6 +565,20 @@ def self.path name
...
+    else; raise TapFormulaAmbiguityError.new(name, p)
```
Comment (jacknagel) [inline Library/Homebrew/formula.rb]: put the raise on a new line here

File: Library/Homebrew/formula.rb [hunk 27]
```diff
@@ -494,7 +494,12 @@ def python(options={}, &block)
 
   # an array of all {Formula} names
   def self.names
-    Dir["#{HOMEBREW_LIBRARY}/Formula/*.rb"].map{ |f| File.basename f, '.rb' }.sort
+    (core_names+tap_names.map { |name| name.split("/")[-1] }.sort).uniq
```
Comment (jacknagel) [inline Library/Homebrew/formula.rb]: spaces around `+`

File: Library/Homebrew/cmd/log.rb (hunk 17 above)
Comment (xu-cheng) [inline Library/Homebrew/cmd/log.rb]: OK, I'm  thinking to rewrite `TapFormulaUnavailableError` as well.

File: Library/Homebrew/formula.rb [hunk 28]
```diff
@@ -494,7 +494,12 @@ def python(options={}, &block)
...
+  end
+
+  # an array of all {Formula} names, which the tap formulae have the fully-qualified name
+  def self.fullnames
```
Comment (jacknagel) [inline Library/Homebrew/formula.rb]: `full_names`?

File: Library/Homebrew/formula.rb [hunk 29]
```diff
@@ -494,7 +494,12 @@ def python(options={}, &block)
...
+    core_names+tap_names
```
Comment (jacknagel) [inline Library/Homebrew/formula.rb]: spaces around `+`

File: Library/Homebrew/exceptions.rb [hunk 30]
```diff
@@ -65,6 +65,22 @@ def to_s; <<-EOS.undent
...
 
+class TapFormulaAmbiguityError < RuntimeError
+  attr_reader :name, :paths
+
+  def initialize name, paths
+    @name = name
+    @paths = paths
+  end
+
+  def to_s; <<-EOS.undent
```
Comment (jacknagel) [inline Library/Homebrew/exceptions.rb]: Instead of redefining `to_s`, just do `super <<-EOS.undent ...` inside initialize

File: Library/Homebrew/formulary.rb (hunk 19 above)
Comment (xu-cheng) [inline Library/Homebrew/formulary.rb]: What I am trying to do is to make sure either we raise error and the formula has been found. Without this line, we may get a state that formula isn't found and won't raise error unless we call `get_formula`. This creates a lot of headache. We have to do a second file checking in the downstream. (There are times that we need check formula existing without the need to read actual formula, like `edit` and `log`). 

So IMHO, this makes things simpler.

**Comment by xu-cheng** [comment]
@jacknagel Updated under your advice.

File: Library/Homebrew/formulary.rb (hunk 19 above)
Comment (xu-cheng) [inline Library/Homebrew/formulary.rb]: @jacknagel Or I can change to let `FormulaLoader.path` be the only place that raises the exception. Is it better?

**Comment by xu-cheng** [comment]
Any further comment or tinker I need to do. 

There are some other refactoring I want to implement. But before I actually works on it. I want to hear the maintainers' advice in case I overdo it. 

* Create a new attribute `full_name` for formula to store full qualified name. Since we drop thr symlink, tapped formula become much more tolerable with the conflict name. Full name become a equal important identity. In the same, it can avoid duplicating string matching. 
* There're `name.downcase` or similar everywhere. I want to put this in the very beginning for performance. 
* Maybe let `core_formula?` become an attribute. It's not useful to compute it every time. 

Also at the meantime, I have some questions.
* Should we enforce to use full qualified name for internal API? We don't need to rush about this as I understand the importance of backward compatibility.
* What should `Fomrula.each` yield? short name as usual or full name?
* What should be used as key in `Fomrulary.klass`? Currently we use short, but will this stop loading the formula with the same short name during `brew readall`?

Would be appropriate for getting answer or any suggestions. :tangerine:

**Comment by jacknagel** [comment]
To set some expectations with regards to the timeline of this pull request, I need to spend some time reviewing the overall design and its impact on other design changes that I'm working on. Given the significance of this change, I would expect it to be at least a few weeks before this is merge-ready.

**Comment by jacknagel** [comment]
>Should we enforce to use full qualified name for internal API? We don't need to rush about this as I understand the importance of backward compatibility.

I'm not sure what exactly this would entail?

>What should Fomrula.each yield? short name as usual or full name?

`Formula.each` yields formula objects.

>What should be used as key in Fomrulary.klass? Currently we use short, but will this stop loading the formula with the same short name during brew readall?

I have some changes pending that load each tap into a separate namespace, which will resolve any issues here.

**Comment by xu-cheng** [comment]
> I would expect it to be at least a few weeks before this is merge-ready

This is well expected for me. In the meaning time, I hope to do some refactoring and optimization of the core as this PR gives me a chance to review part of internal code.

> I have some changes pending that load each tap into a separate namespace, which will resolve any issues here.

:+1:

**Comment by xu-cheng** [comment]
There's still a long way before this PR can be merged. So I will spilt some unrelated code refactors to separate PRs.

@mikemcquaid I just found out [Homebrew task on Google Summer of Code](https://github.com/github/gsoc#allow-prioritisation-of-taps). Doesn't this PR solve the task Allow prioritisation of taps? :stuck_out_tongue_winking_eye:

**Comment by mikemcquaid** [comment]
@xu-cheng Yeh, that'd be good thanks. I don't think it does solve it but it does make the solution easier. By prioritisation I'm thinking allowing a way to make e.g. `brew install wget` install from a tap rather than Homebrew core if you've configured e.g. an environment variable to do that.

**Comment by xu-cheng** [comment]
Close in favor of #39524
